src/client_metrics.rs

```diff
@@ -171,105 +173,195 @@ pub struct MetricsMetadata {
+        Some(self.cmp(other))
     }
 }
 
-impl Eq for MetricSample {}
+impl Ord for Bucket {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.le
+            .partial_cmp(&other.le)
+            .unwrap_or(std::cmp::Ordering::Equal)
+    }
+}
 
-impl MetricSample {
-    pub fn labels_to_key(&self) -> String {
-        match &self.labels {
-            Some(labels_map) => {
-                let mut sorted_entries: Vec<(&String, &String)> = labels_map.iter().collect();
-                sorted_entries.sort_unstable_by(|a, b| a.0.cmp(b.0));
+fn deserialize_bucket_le<'de, D>(deserializer: D) -> Result<f64, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    use serde::de::Error;
+
+    #[derive(Deserialize)]
+    #[serde(untagged)]
+    enum BucketLe {
+        Number(f64),
+        String(String),
+    }
 
-                let mut key = String::with_capacity(sorted_entries.len() * 16);
+    match BucketLe::deserialize(deserializer)? {
+        BucketLe::Number(n) => Ok(n),
+        BucketLe::String(s) if s == "+Inf" => Ok(f64::INFINITY),
+        BucketLe::String(s) => Err(D::Error::custom(format!("expected '+Inf', got '{}'", s))),
+    }
+}
 
-                for (i, (k, v)) in sorted_entries.iter().enumerate() {
-                    if i > 0 {
-                        key.push(',');
-                    }
-                    key.push_str(k);
-                    key.push(':');
-                    key.push_str(v);
-                }
-                key
-            }
-            None => String::new(),
-        }
+fn serialize_bucket_le<S>(le: &f64, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    if le.is_infinite() {
+        serializer.serialize_str("+Inf")
+    } else {
+        serializer.serialize_f64(*le)
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[derive(Debug, Clone, Deserialize, Serialize)]
 #[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
-#[serde(rename_all = "lowercase")]
-pub enum MetricType {
-    Counter,
-    Gauge,
-    #[serde(other)]
-    Unknown,
+#[serde(rename_all = "camelCase")]
+pub struct BucketMetricSample {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub labels: MetricLabels,
+    pub count: u64,
+    pub sum: f64,
+    pub buckets: Vec<Bucket>,
+}
+
+impl PartialEq for BucketMetricSample {
+    fn eq(&self, other: &Self) -> bool {
+        self.labels == other.labels
+            && self.count == other.count
+            && (self.sum - other.sum).abs() < f64::EPSILON
+            && self.buckets == other.buckets
+    }
 }
 
-impl FromStr for MetricType {
-    type Err = std::convert::Infallible;
+impl Eq for BucketMetricSample {}
+
+impl MergeMut for BucketMetricSample {
+    fn merge(&mut self, other: BucketMetricSample) {
+        self.count += other.count;
+        self.sum += other.sum;
+
+        // Optimized for the common case where both samples have the same bucket boundaries
```

Comment:
also tested in benchmarks

Reply:
First rule of optimization is don't do it until you have a benchmark. Second rule of optimization is check that your optimized code is actually faster. With the fast path just commented out, your benchmark produces a 19% improvement in speed on my machine. Std deviation is probably a little high but the optimized code seems roughly as slow or slower, can we yeet it? It's Friday so I'm not checking flame charts right now but I suspect the memory is the same.

Sanity check me? Did you expect your benchmark to give a perf bump with the fast path executing?

Reply:
yes, it was premature optimization that actually degraded performance. Double checked on my machine and you're right. Removing the "fast path" which was actually slower